### PR TITLE
Deserialize the payloads when deserializing `Message`s

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by erikzhang** [comment]
@shargon We need to move forward quickly.

File: neo/Network/P2P/ProtocolHandler.cs [hunk 1]
```diff
@@ -57,53 +57,53 @@ protected override void OnReceive(object message)
...
+                    OnPingMessageReceived((PingPayload)msg.Payload);
                     break;
                 case MessageCommand.Pong:
-                    OnPongMessageReceived(msg.GetPayload<PingPayload>());
+                    OnPongMessageReceived((PingPayload)msg.Payload);
                     break;
                 case MessageCommand.Transaction:
-                    if (msg.Payload.Length <= Transaction.MaxTransactionSize)
-                        OnInventoryReceived(msg.GetTransaction());
+                    if (msg.Payload.Size <= Transaction.MaxTransactionSize)
```
Comment (shargon) [inline neo/Network/P2P/ProtocolHandler.cs:105]: This is slow again, we should expose the Size in a property maybe

File: neo/Network/P2P/Message.cs [hunk 2]
```diff
@@ -105,26 +97,61 @@ public static int TryDeserialize(ByteString data, out Message msg)
             {
                 Flags = flags,
                 Command = (MessageCommand)header[1],
-                Payload = data.Slice(payloadIndex, (int)length).ToArray()
+                _payload_compressed = data.Slice(payloadIndex, (int)length).ToArray()
             };
+            msg.SetPayload();
```
Comment (shargon) [inline neo/Network/P2P/Message.cs]: Maybe we just want to relay the message, we should set the payload when we call the property

Comment (shargon) [reply neo/Network/P2P/Message.cs]: Fixed in https://github.com/shargon/neo/pull/9/commits/f22fa21bfce22e4396ee00b84b099ead848a448b

File: neo/Network/P2P/ProtocolHandler.cs (hunk 1 above)
Comment (shargon) [reply neo/Network/P2P/ProtocolHandler.cs:105]: The fix for this is more dirty... `msg.PayloadSize`?

File: neo/Network/P2P/Message.cs (hunk 2 above)
Comment (erikzhang) [reply neo/Network/P2P/Message.cs]: We always access `Payload` whether or not we relay the message. For example, we must verify tx before relay it. So we decompress the payload when deserializing the message to check the real size of payload.

File: neo/Network/P2P/ProtocolHandler.cs (hunk 1 above)
Comment (erikzhang) [reply neo/Network/P2P/ProtocolHandler.cs:105]: `Transaction.Size` is slow, but not too slow. In NEO 3.0, we don't have utxo in transactions, so the size of transactions are easy to be calculated.

Comment (shargon) [reply neo/Network/P2P/ProtocolHandler.cs:105]: Is not better to store the value?

Comment (erikzhang) [reply neo/Network/P2P/ProtocolHandler.cs:105]: If we decide to cache the size value, we can cache it in `Transaction`.

**Review by shargon — APPROVED** [review]
The PR is good, just the recommendation about store the Size of the payload. But `Transaction` could do that in a future

**Pull request merged**